client/src/Components/Session/withAuthentication.js

```diff
@@ -0,0 +1,60 @@
+/*
+ * Higher order component which makes token and user object available
+ * to all components in Context. Components can also update token and
+ * userId via setToken() and setUserId() methods
+ */
+
+import React, { useEffect, useState } from "react";
+import { getUser } from "Utils/api";
+
+import AuthUserContext from "./AuthUserContext";
+
+const withAuthentication = Component => {
+  const WithAuthentication = props => {
+    const prevToken = window.localStorage.getItem("token") || null;
+    const prevUserId = window.localStorage.getItem("userId") || null;
+
+    const [token, setToken] = useState(prevToken);
+    const [userId, setUserId] = useState(prevUserId);
+    const [user, setUser] = useState({});
+
+    // When token or userId changes via setToken or setUserId,
+    // save to local storage, fetch the user and put in context,
+    // or delete if logged out
+    useEffect(() => {
+      const getCurrentUser = async () => {
+        if (token) {
+          window.localStorage.setItem("token", token);
+        } else {
+          window.localStorage.removeItem("token");
```

Comment:
If this is the case, do we even need to make the API call, since the user would be logged out right?

Reply:
Not sure I understand fully, but I'm also not sure I'm doing this in the most optimal way.

A log out function would call `setToken("null")` and `setUserId("null")`, which as far as I understand this hook, triggers this function twice. Once on `token` change and once on `userId` change. Which is why I need these `if()` checks. However, the first time it runs on `setToken("null")`, it will probably make an unnecessary API call, because userId is still set.

Having thought about it, I think the solution is to not have the userId in localStorage at all. We don't need it to get the user object (I initially thought we did) and we can store it in the context.user object. Is it normal to just have the token in storage?

Reply:
The userId is already embedded in the jwt token (since it was signed with the userId as the payload). Also, the getUser route already returns an object with the userId, so it's redundant to store it in localStorage

Reply:
Thanks Danny. I've removed userId from state and `api/users` calls, which should fix the problem here and make the code cleaner:

```
  useEffect(() => {
      const getCurrentUser = async () => {
        if (token) {
          window.localStorage.setItem("token", token);
          axios.defaults.headers.common["Authorization"] = token;
          const loggedInUser = await getUser();
          setUser(loggedInUser);
        }
      };
      getCurrentUser();
    }, [token]);
```